Interfaces: Expose UTXO Snapshot Loading and Add Progress Notifications

D33r-Gee opened the pull request:
# Add UTXO Snapshot Loading Interface and Progress Notifications

This PR implements a complete interface for loading UTXO snapshots through the GUI, including progress notifications to provide user feedback during the loading process.

---

## Changes

### Commit 1: interfaces: expose UTXO snapshot functionality (loading/activate) in node interface

- Added a new virtual method `snapshot` to the node interface for loading UTXO snapshots from a path.
- Added `snapshot.h` to `src/interfaces` defining the Snapshot interface; `activate()` returns `util::Result<const CBlockIndex*>` to   carry either the base block index on success or a descriptive error message on failure, matching the shape of `ChainstateManager::ActivateSnapshot()`.
- Added a `getMetadata` accessor so callers can read snapshot metadata populated during activation without re-parsing the file.
- Added `SnapshotImpl` in src/node/interfaces.cpp: opens the snapshot file, parses metadata, and forwards to `ChainstateManager::ActivateSnapshot()`, propagating any error as util::Error{...}.

This change enhances the user's ability to load UTXO snapshots through the GUI and/or RPC.

### Commit 2: `notifications: Implement snapshot load progress notifications`

- Added a new signal `SnapshotLoadProgress` for snapshot load progress in the UI interface.
- Updated the notifications interface to include a method for reporting snapshot loading progress.
- Enhanced the `ChainstateManager` to report progress during snapshot loading.
- Implemented the necessary handler in the kernel notifications to relay progress updates.

This change improves user feedback during the snapshot loading process.

### Commit 3: `rpc: refactor loadtxoutset to use snapshot interface`

- Updated the `loadtxoutset` RPC to call the new snapshot interface instead of directly working with chainstate internals. File opening, metadata parsing, and activation now all happen inside `interfaces::Snapshot::activate()` rather than being split between the RPC handler and the interface.
- Removed the redundant pre-parse block in the RPC handler; its file-open and metadata-parse work duplicated what activate() already performs.
- Collapsed the RPC error handling to a single `RPC_INTERNAL_ERROR` built from `util::ErrorString(activation_result).original`, using the `"Unable to load UTXO snapshot: %s. (%s)"` format established in #30267 / #30395.
- Read `coins_loaded` from `snapshot->getMetadata()` after `activate()` populates it.
- Updated `feature_assumeutxo.py` to assert on the unified error code for file-open and metadata-parse failures (message substrings unchanged).
Improves modularity by consolidating snapshot handling through the interface layer.
---

## Technical Details

The implementation includes:

- **UI Interface Enhancement**
  Added `SnapshotLoadProgress` signal to `CClientUIInterface` with corresponding signal declaration and implementation.

- **Node Interface**
  Added `loadSnapshot` virtual method to the node interface with implementation in `interfaces.cpp`.

- **Progress Notifications**
  Implemented `snapshotLoadProgress` method in `KernelNotifications` class to relay progress updates to the UI.

- **Base Interface**
  Added `snapshotLoadProgress` virtual method to the kernel notifications interface.

---

## Benefits

- **User Experience**
  Provides real-time feedback during UTXO snapshot loading operations.

- **GUI Integration**
  Enables loading UTXO snapshots directly through the Bitcoin Core GUI.

- **Progress Tracking**
  Users can monitor the progress of snapshot loading operations.

- **Consistent Interface**
  Follows the existing pattern for progress notifications in the codebase.

## Testing

- **Daemon** build and launch `bitcoind` then with `bitcoin-cli` load a snapshot using:
```bash
./bitcoin-cli -rpcclienttimeout=0 -signet loadtxoutset /path/to/utxo-snapsshot.dat
```
- **QT Widgets**: build and run [this](https://github.com/bitcoin-core/gui/pull/870) draft PR in legacy gui repo 
- **QML** version: for a modern look build and run [this](https://github.com/bitcoin-core/gui-qml/pull/488) PR in the newly updated gui-qml repo

D33r-Gee committed:
interfaces: expose UTXO snapshot functionality (loading/activate) in node interface

- Added a new virtual method `snapshot` to the node interface for loading UTXO snapshots.
- Added `snapshot.h` to src/interfaces defining the Snapshot interface. `activate()`
  returns `util::Result<const CBlockIndex*>` to carry either the base block index
  on success or a descriptive error message on failure, matching the shape of
  `ChainstateManager::ActivateSnapshot()`.
- Added a `getMetadata` accessor so callers can read snapshot metadata populated
  during activation without re-parsing the file.
- Added `SnapshotImpl` in src/node/interfaces.cpp: opens the snapshot file,
  parses metadata, and forwards to `ChainstateManager::ActivateSnapshot()`,
  propagating any error as `util::Error{...}`.

This change enhances the user ability to load utxo snapshots through the GUI and/or RPC.

D33r-Gee committed:
notifications: Implement snapshot load progress notifications

- Added a new signal for snapshot load progress in the UI interface.
- Updated the notifications interface to include a method for reporting snapshot loading progress.
- Enhanced the ChainstateManager to report progress during snapshot loading.
- Implemented the necessary handler in the kernel notifications to relay progress updates.

This change improves user feedback during the snapshot loading process.

D33r-Gee commented:
> I plan to test this soon(tm).

Thanks @Sjors !

ryanofsky commented on src/interfaces/node.h:
```diff
@@ -204,6 +205,9 @@ class Node
     virtual std::vector<std::string> listRpcCommands() = 0;
 
+    //! Load UTXO Snapshot.
+    virtual bool loadSnapshot(AutoFile& afile, const node::SnapshotMetadata& metadata, bool in_memory) = 0;
```
In commit "interfaces: expose UTXO snapshot loading in node interface" (de11e3b89f8aba9a90a3419086807e5fee0f941d)

Note with #10102 if the GUI & node are running in separate processes, it could be a awkward for the gui process to create an AutoFile object and pass it to the node like this. Technically it could probably work: the AutoFile could be passed as a file descriptor or a (path, position) tuple. But it doesn't seem like a clean separation of responsibilities if the GUI process is reading the beginning of the snapshot file and the node process is reading the rest of the file.

Probably current interface is ok, but it might be more ideal if node provided a more generic `loadSnapshot` method and the GUI didn't directly access the file:

```c++
class Node
{
    // ...
    std::unique_ptr<Snapshot> loadSnapshot(const fs::path& path) = 0;
    // ...
};

class Snapshot
{
public:
    virtual ~Snapshot() = default;
    node::SnapshotMetadata getMetadata() = 0;
    bool activate() = 0;
};
```

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/33117#discussion_r2249012242

Thanks! I think b464f388197c5c31937d309d817d38c4b4849a91 is an improvement over previous de11e3b89f8aba9a90a3419086807e5fee0f941d because it passes an `fs::path` argument from the GUI to the node instead of an `AutoFile` file argument. This is nice because it lets node be fully responsible for parsing the snapshot file instead of relying on the GUI.

One comment on b464f388197c5c31937d309d817d38c4b4849a91 is that it looks like the Snapshot class is not actually necessary in its current form. The reason I suggested adding a Snapshot class above was that I thought GUI wanted to display SnapshotMetadata information before activating the snapshot, and a Snapshot class could provide separate getMetadata() and activate() methods to allow that. But since the Snapshot class only has a single activate() method in b464f388197c5c31937d309d817d38c4b4849a91, it looks like the class is unnecessary, and you could just have a plain `Node::activateSnapshot(fs::path)` method that returns bool instead of a Snapshot.

Any approach is fine though, so you should choose whatever seems to make the most sense.

pinheadmz commented:
concept ACK, tested the mainnet snapshot load with the QML gui branch built on top of this

D33r-Gee commented:
with [8bc4035](https://github.com/bitcoin/bitcoin/commit/8bc4035570acfe512c756c6de30d6440d322d514) it has been rebased and the fuzzing CI error has been addressed

Sjors reviewed (commented):
At first glance this looks very reasonable. I also tested it in the (current) GUI with https://github.com/bitcoin-core/gui/pull/870

I had a similar concern as @ryanofsky about how to go about passing the snapshot file (path) over the interface. The current approach in 3c8b578836dd0d8438a8aac81a1166df84811562 is the same as how we handle wallet backups (`restoreWallet`), so probably fine?

TheCharlatan commented:
Approach ACK

D33r-Gee committed:
rpc: Refactor loadtxoutset to utilize new node interface for UTXO snapshot loading

- Updated the loadtxoutset function to use the new node interface for loading
  UTXO snapshots. File opening, metadata parsing, and activation now all happen
  inside `interfaces::Snapshot::activate()` rather than being split between the
  RPC handler and the interface.
- Replaced direct file handling with snapshot activation through the node
  interface. The previous pre-parse block in the RPC handler was redundant
  with the equivalent work already performed inside `activate()`.
- Collapsed the RPC error handling to a single `RPC_INTERNAL_ERROR` built from
  `util::ErrorString(activation_result).original`, matching the convention
  established in #30267 / #30395.
- Enhanced the result object to correctly reference metadata from the snapshot
  via `snapshot->getMetadata()`.
- Updated `feature_assumeutxo.py` to assert on the unified error code for
  file-open and metadata-parse failures.

This change streamlines the UTXO snapshot loading process and leverages the new interface capabilities.

pinheadmz reviewed (commented):
code review ACK at  d0c32c28e5

The new interface and signal handlers are fairly straight-forward and follow existing patterns. Consuming the new interface in the `loadtxoutset` RPC ensures it is covered by tests. Corecheck indicates some of the failure cases are not covered by tests, I want to look at that a little closer. 

I also plan on going through the flow with both GUIs as well. I think this is an important feature to expose to GUI users, and this is the right way to get there.

Only one question below so far

D33r-Gee commented:
With [8c932ff](https://github.com/bitcoin/bitcoin/commit/8c932ff59f58ee2f39a58cbeeaf8c16722c8c272) addressed @pinheadmz comment about the loading progress notifications also it has been rebased over master

pinheadmz commented on src/interfaces/snapshot.h:
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2024 The Bitcoin Core developers
```
740add96ec6149174d69b05cab1088dd7f7dc776

nit: can remove the year
`Copyright (c) The Bitcoin Core developers`

D33r-Gee replied:
Done...

pinheadmz commented on src/node/interfaces.cpp:
```diff
@@ -94,6 +97,73 @@ namespace node {
+    std::optional<const CBlockIndex*> getActivationResult() const override
+    {
+        if (m_activation_result.has_value()) {
+            return m_activation_result;
+        }
+        return std::nullopt;
+    }
```
740add96ec6149174d69b05cab1088dd7f7dc776

Since `m_activation_result` is already an optional I think you can shortcut the logic here?

```cpp
std::optional<const CBlockIndex*> getActivationResult() const override
{
    return m_activation_result;
}
```

D33r-Gee replied:
Addressed in a later refactor([328fef6](https://github.com/bitcoin/bitcoin/commit/328fef6bae4007cb3903c6d2b5653f91c9bf14be)): removed getActivationResult() and the m_activation_result member. activate() now returns util::Result<const CBlockIndex*>, so callers get success/failure and the block index from one call without a separate optional getter.

pinheadmz commented on src/rpc/blockchain.cpp:
```diff
@@ -3408,27 +3410,37 @@ static RPCMethod loadtxoutset()
+        }
+        if (error_msg.find("Snapshot file does not exist") != std::string::npos) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Couldn't open file " + path.utf8string() + " for reading.");
+        }
```
9dbfdf81caff7cab40db94be9553601eacffd63c

Seems a little awkward to parse error strings for the sake of RPC errors. I think the generic one you have below on L3431 is fine, and if you want to include the failed path in the error message, it could happen over in `class SnapshotImpl` where the strings are set.

D33r-Gee replied:
Fixed in af0c253: dropped getLastError() and string matching. loadtxoutset only calls activate() and returns a single RPC_INTERNAL_ERROR with util::ErrorString(activation_result).original (same pattern as #30267 / #30395). Path-specific messages come from SnapshotImpl::activate().

pinheadmz commented on src/rpc/blockchain.cpp:
```diff
@@ -3408,27 +3410,37 @@ static RPCMethod loadtxoutset()
+    const node::SnapshotMetadata* metadata;
     try {
-        afile >> metadata;
+        metadata = &snapshot->getMetadata();
```
9dbfdf81caff7cab40db94be9553601eacffd63c

Would this ever throw? Probably don't need the try/catch

D33r-Gee replied:
Yes! That duplication is gone in af0c253. File open, metadata parse, and activation all live in SnapshotImpl::activate(). The RPC only does snapshot(path) → activate() → build the result from getMetadata() on success.

pinheadmz commented on src/rpc/blockchain.cpp:
```diff
@@ -3408,27 +3410,37 @@ static RPCMethod loadtxoutset()
 
-    auto activation_result{chainman.ActivateSnapshot(afile, metadata, false)};
+    auto activation_result = snapshot->getActivationResult();
     if (!activation_result) {
```
9dbfdf81caff7cab40db94be9553601eacffd63c

The snapshot is activated (or not) up on line 3419. Will this call to `getActivationResult()` ever be useful after that?

D33r-Gee replied:
Agreed! Removed in af0c253. getMetadata() is used directly after a successful activate(); no try/catch around it.

pinheadmz reviewed (commented):
code review 9dbfdf81ca

h/t Claude !

D33r-Gee commented:
with [9dab30a](https://github.com/bitcoin/bitcoin/commit/9dab30ad3d82567e60a18d7a657f0420d604b04f) addressed @pinheadmz comments. Thanks for reviewing!

pinheadmz commented on src/rpc/blockchain.cpp:
```diff
@@ -3408,37 +3410,40 @@ static RPCMethod loadtxoutset()
-            RPC_INVALID_PARAMETER,
-            "Couldn't open file " + path.utf8string() + " for reading.");
+    auto node_interface = interfaces::MakeNode(node);
+    node::SnapshotMetadata metadata{EnsureAnyChainman(request.context).GetParams().MessageStart()};
+    {
+        FILE* file{fsbridge::fopen(path, "rb")};
+        AutoFile afile{file};
+        if (afile.IsNull()) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Couldn't open file " + path.utf8string() + " for reading.");
+        }
+        try {
+            afile >> metadata;
+        } catch (const std::ios_base::failure& e) {
+            throw JSONRPCError(RPC_DESERIALIZATION_ERROR, strprintf("Unable to parse metadata: %s", e.what()));
+        }
```
I don't understand whats happening here. If the file can't be opened for reading, or metadata can't be parsed, won't that come up in `activate()` on line 3433?

D33r-Gee replied:
> I don't understand whats happening here. If the file can't be opened for reading, or metadata can't be parsed, won't that come up in `activate()` on line 3433?

Thanks for reviewing @pinheadmz and good catch! This led me to revise the approach with the help of Claude.

You are correct activate() already opens the file and parses metadata internally, so doing it again here was redundant. The only reason these specific RPC_INVALID_PARAMETER / RPC_DESERIALIZATION_ERROR codes were being produced at the RPC layer was to retrofit finer error classification on top of the old bool activate() signature.

D33r-Gee replied:
Agreed! Removed in af0c253. activate() returns util::Result<const CBlockIndex*>; the RPC uses that result directly and no longer has getActivationResult() or m_activation_result.

fanquake commented:
Are you still working on this? You should try and respond to review comments inline, so it's clear what has been done/responded to, and what hasn't.

D33r-Gee commented:
> Are you still working on this? You should try and respond to review comments inline, so it's clear what has been done/responded to, and what hasn't.

Yes! Still working on it with [af0c253](https://github.com/bitcoin/bitcoin/commit/af0c25311f084217e43502dd8f7081b7a77ab44a) it has been rebased over master.

Will take some time to add comments inline like you suggested @fanquake

D33r-Gee commented on src/validation.cpp:
```diff
@@ -5963,6 +5963,10 @@ util::Result<void> ChainstateManager::PopulateAndValidateSnapshot(
 
+                // Show Snapshot Loading progress
+                double progress = static_cast<double>(coins_processed) / static_cast<double>(coins_count);
+                GetNotifications().snapshotLoadProgress(progress);
```
Still the case on af0c253

sedited commented:
@D33r-Gee can you rebase this?

D33r-Gee commented:
> @D33r-Gee can you rebase this?

Yep... Will do so this week...

D33r-Gee commented:
with [5142e2e](https://github.com/bitcoin/bitcoin/commit/5142e2e8fe1c22c6bcc095854b80afc389ead55a) it has been rebased over master.

Thanks @sedited for the nudge :)

D33r-Gee commented:
With [486e6d7](https://github.com/bitcoin/bitcoin/commit/486e6d72000f90eb2431d026dc639a7b3169872c) it has been rebased over master

D33r-Gee commented:
Updated with [c765d9f](https://github.com/bitcoin/bitcoin/commit/c765d9fa113f032c0f6fa54369e3add1a7edcb38) it has been rebased over master
